### ISSUE TITLE
Move to forbidden fails

### ISSUE THREAD

**Issue by donalgrant** [donalgrant/p6-equations#12]
See output:
```
Player 1:
***I'm crazily moving - to forbidden
Player 2:
creating a bag from missing=2; allowed=2 1 7 7 ^ ^ - - 3 3 3 * +; rpn=2 7 7 ^ - 3 3 * +
bag sums are:  missing+allowed=3 3 3 * + 1 2 2 ^ ^ - - 7 7; rpn-missing=^ - 7 7 3 3 * +
Setting up a Board with cubes 2 2 1 ^ - 3, which should include missing cube 2
Board from BagHash 2,2,1,^,-,3
  ________________________________________                            
  Unused:     - ^ 1 2 2 3 (6 cubes)
  Required:    (0 cubes)
  Permitted:   (0 cubes)
  Forbidden:   (0 cubes)
  Goal:       
  ________________________________________

Board from BagHash 2,2,1,^,-,3
calculate_solutions for ncubes=3
***I'm moving 2 to forbidden (because I can replace it in 23^7+73-*)
2 not available to move  in code  at SETTING::src/core/Backtrace.pm6 line 85
   in method new at SETTING::src/core/Backtrace.pm6 line 85
   in sub caller at /Users/imel/gitdev/donalgrant/p6-equations/lib/Globals.pm6 line 7
   in sub msg at /Users/imel/gitdev/donalgrant/p6-equations/lib/Globals.pm6 line 12
   in sub quit at /Users/imel/gitdev/donalgrant/p6-equations/lib/Globals.pm6 line 16
   in method move at /Users/imel/gitdev/donalgrant/p6-equations/lib/Board.pm6 line 53
   in method move_to_forbidden at /Users/imel/gitdev/donalgrant/p6-equations/lib/Board.pm6 line 60
   in block  at /Users/imel/gitdev/donalgrant/p6-equations/lib/Player.pm6 line 130
   in block  at /Users/imel/gitdev/donalgrant/p6-equations/lib/Player.pm6 line 126
   in method turn at /Users/imel/gitdev/donalgrant/p6-equations/lib/Player.pm6 line 104
   in block <unit> at equations.p6 line 49

Died
  in sub quit at /Users/imel/gitdev/donalgrant/p6-equations/lib/Globals.pm6 line 16
  in method move at /Users/imel/gitdev/donalgrant/p6-equations/lib/Board.pm6 line 53
  in method move_to_forbidden at /Users/imel/gitdev/donalgrant/p6-equations/lib/Board.pm6 line 60
  in block  at /Users/imel/gitdev/donalgrant/p6-equations/lib/Player.pm6 line 130
  in method turn at /Users/imel/gitdev/donalgrant/p6-equations/lib/Player.pm6 line 104
  in block <unit> at equations.p6 line 49

northwind:p6-equations imel$ 
```

**Comment by donalgrant** [comment]
This was happening during the use of find_replacement to extend doable solutions, rather than not_doable solutions.  In some cases, the replacement cube was already on the board in the permitted or required section, and was not in unused.  This means that the move to forbidden is not available.  Put in a new check to only make the move to forbidden if the replacement cube is unused.  

This brings up a larger topic about the logic of these moves:  might only be worth doing the forbidden move if the move actually disables that particular solution.  So could check to make sure that removing one would disable the equation before doing the forbidden move.  (Ties possibly into the bonus move, as well.)  

Additional logic:  better than a forbidden move might be a required move for the extended equation, if it brings in an operator or digit not in the original rpn.

Once the move is made, and before the method returns, need to filter the solutions based on the newly forbidden cube.  (Previously, the find_replacement had to be called by the player again at the next turn when some RPNs were no longer "doable".)  This ties into a separate issue to put all the move-related action into a single method, so that the filtering would happen there as well.

Finally, in the process of debugging this, I realized that there might be a need for a Board.clone method to make copies of Boards which can be altered without changing the original.  Created that method, and added tests to Board.t6.

**Comment by donalgrant** [comment]
Another issue causing this error is that a number d can get replaced with the equation '0+d', which looks like a replacement equation, but then fails when the only 'd' available gets moved to forbidden.

**Comment by donalgrant** [comment]
Refactored find_replacement to return Nil on failure and an array of rpn-strings on success.  No longer changes Player list of solutions -- that will be done (if desired) by the calling method.

**Comment by donalgrant** [comment]
Turns out interpreting Nil as an array is bad.  Changed find_replacement to return an empty array if none were found.

**Comment by donalgrant** [comment]
Fixed by commit 31286c500abec3d0fa0e687c190654b2f0ca4b12
